File: go-backend-interview-prep/10_design_patterns_in_go/01_constructor_and_di/main.go
package main

import "fmt"

// --- Interfaces (dependencies) ---

type Sender interface {
	Send(to, message string) error
}

type Logger interface {
	Log(msg string)
}

// --- Concrete implementations ---

type EmailSender struct{}

func (e *EmailSender) Send(to, message string) error {
	fmt.Printf("  [EMAIL] To: %s | Message: %s\n", to, message)
	return nil
}

type SMSSender struct{}

func (s *SMSSender) Send(to, message string) error {
	fmt.Printf("  [SMS]   To: %s | Message: %s\n", to, message)
	return nil
}

type ConsoleLogger struct{}

func (c *ConsoleLogger) Log(msg string) {
	fmt.Printf("  [LOG]   %s\n", msg)
}

// --- Service (depends on interfaces, not concrete types) ---

type NotificationService struct {
	sender Sender
	logger Logger
}

// NewNotificationService is the constructor -- dependencies are injected here.
func NewNotificationService(s Sender, l Logger) *NotificationService {
	return &NotificationService{sender: s, logger: l}
}

func (n *NotificationService) Notify(to, message string) error {
	n.logger.Log("sending notification to " + to)
	return n.sender.Send(to, message)
}

// --- Fake for testing ---

type FakeSender struct {
	Calls []string
}

func (f *FakeSender) Send(to, message string) error {
	f.Calls = append(f.Calls, to+":"+message)
	return nil
}

func main() {
	fmt.Println("=== Constructor and Dependency Injection ===")
	fmt.Println()

	logger := &ConsoleLogger{}

	// Wire with email sender
	fmt.Println("--- Email notification ---")
	emailSvc := NewNotificationService(&EmailSender{}, logger)
	emailSvc.Notify("alice@example.com", "Your order shipped!")
	fmt.Println()

	// Wire with SMS sender -- same service, different behavior
	fmt.Println("--- SMS notification ---")
	smsSvc := NewNotificationService(&SMSSender{}, logger)
	smsSvc.Notify("+1234567890", "Your order shipped!")
	fmt.Println()

	// Wire with fake for testing
	fmt.Println("--- Test with fake ---")
	fake := &FakeSender{}
	testSvc := NewNotificationService(fake, logger)
	testSvc.Notify("bob@test.com", "test message")
	fmt.Printf("  Fake recorded %d call(s): %v\n", len(fake.Calls), fake.Calls)
	fmt.Println()

	fmt.Println("Key: same service, different senders. DI makes it testable.")
}

File: go-backend-interview-prep/10_design_patterns_in_go/01_constructor_and_di/main.py
"""Constructor and dependency injection -- Python equivalent of the Go example.

Uses duck typing for dependencies (no explicit interface needed).
"""


# --- Concrete implementations ---

class EmailSender:
    def send(self, to: str, message: str) -> None:
        print(f"  [EMAIL] To: {to} | Message: {message}")


class SMSSender:
    def send(self, to: str, message: str) -> None:
        print(f"  [SMS]   To: {to} | Message: {message}")


class ConsoleLogger:
    def log(self, msg: str) -> None:
        print(f"  [LOG]   {msg}")


# --- Service (depends on duck-typed sender and logger) ---

class NotificationService:
    def __init__(self, sender, logger):
        """Constructor injection: dependencies come in as arguments."""
        self.sender = sender
        self.logger = logger

    def notify(self, to: str, message: str) -> None:
        self.logger.log(f"sending notification to {to}")
        self.sender.send(to, message)


# --- Fake for testing ---

class FakeSender:
    def __init__(self):
        self.calls: list[str] = []

    def send(self, to: str, message: str) -> None:
        self.calls.append(f"{to}:{message}")


def main() -> None:
    print("=== Constructor and Dependency Injection ===")
    print()

    logger = ConsoleLogger()

    # Wire with email sender
    print("--- Email notification ---")
    email_svc = NotificationService(EmailSender(), logger)
    email_svc.notify("alice@example.com", "Your order shipped!")
    print()

    # Wire with SMS sender
    print("--- SMS notification ---")
    sms_svc = NotificationService(SMSSender(), logger)
    sms_svc.notify("+1234567890", "Your order shipped!")
    print()

    # Wire with fake for testing
    print("--- Test with fake ---")
    fake = FakeSender()
    test_svc = NotificationService(fake, logger)
    test_svc.notify("bob@test.com", "test message")
    print(f"  Fake recorded {len(fake.calls)} call(s): {fake.calls}")
    print()

    print("Key: same service, different senders. DI makes it testable.")


if __name__ == "__main__":
    main()
